File: json/json_test.go
package json

import (
	"testing"
	"math"
	"fmt"
)

func isDoubleEqual(f1, f2 float64) bool {
	const TOLERANCE = 0.000001
	return math.Abs(f1-f2) < TOLERANCE
}

func checkNumber(v float64, ctx *Context, t *testing.T) {
	node, err := ctx.ParseNumber()
	if err != nil || !isDoubleEqual(node, v) {
		t.Fatal(fmt.Sprintf("err : %v", ctx.json))
	}
}

func TestNumber(t *testing.T) {
	checkNumber(1.0, &Context{json: "1 "}, t)
	checkNumber(0.0, &Context{json: "-0 "}, t)
	checkNumber(0.0, &Context{json: "-0.0 "}, t)
	checkNumber(1, &Context{json: "1 "}, t)

	checkNumber(-1.0, &Context{json: "-1 "}, t)
	checkNumber(1.5, &Context{json: "1.5 "}, t)
	checkNumber(-1.5, &Context{json: "-1.5 "}, t)
	checkNumber(3.1416, &Context{json: "3.1416 "}, t)
	checkNumber(1E10, &Context{json: "1E10 "}, t)
	checkNumber(1e10, &Context{json: "1e10 "}, t)
	checkNumber(1E+10, &Context{json: "1E+10 "}, t)
	checkNumber(1E-10, &Context{json: "1E-10 "}, t)
	checkNumber(-1E10, &Context{json: "-1E10 "}, t)
	checkNumber(-1e10, &Context{json: "-1e10 "}, t)
	checkNumber(-1E+10, &Context{json: "-1E+10 "}, t)
	checkNumber(-1E-10, &Context{json: "-1E-10 "}, t)
	checkNumber(1.234E+10, &Context{json: "1.234E+10 "}, t)
	checkNumber(1.234E-10, &Context{json: "1.234E-10 "}, t)
	checkNumber(0.0, &Context{json: "1e-10000 "}, t) /* must underflow */

	checkNumber(1.0000000000000002, &Context{json: "1.0000000000000002 "}, t)           /* the smallest number > 1 */
	checkNumber(4.9406564584124654e-324, &Context{json: "4.9406564584124654e-324 "}, t) /* minimum denormal */
	checkNumber(-4.9406564584124654e-324, &Context{json: "-4.9406564584124654e-324 "}, t)
	checkNumber(2.2250738585072009e-308, &Context{json: "2.2250738585072009e-308 "}, t) /* Max subnormal double */
	checkNumber(-2.2250738585072009e-308, &Context{json: "-2.2250738585072009e-308 "}, t)
	checkNumber(2.2250738585072014e-308, &Context{json: "2.2250738585072014e-308 "}, t) /* Min normal positive double */
	checkNumber(-2.2250738585072014e-308, &Context{json: "-2.2250738585072014e-308 "}, t)
	checkNumber(1.7976931348623157e+308, &Context{json: "1.7976931348623157e+308 "}, t) /* Max double */
	checkNumber(-1.7976931348623157e+308, &Context{json: "-1.7976931348623157e+308 "}, t)

}

func checkString(v string, ctx *Context, t *testing.T) {
	value, err := ctx.ParseString()
	if err != nil || value != v {
		t.Fatal(fmt.Sprintf("err : %v", ctx.json))
	}
}

func TestString(t *testing.T) {
	checkString("", &Context{json: `"" `}, t)
	checkString("Hello", &Context{json: `"Hello" `}, t)
	checkString("Hello\nWorld", &Context{json: `"Hello\nWorld" `}, t)
	checkString("\" \\ / \b \f \n \r \t", &Context{json: `"\" \\ \/ \b \f \n \r \t" `}, t)
	checkString("Hello\u0000World", &Context{json: "\"Hello\\u0000World\" "}, t)
	checkString("\x24", &Context{json: `"\u0024" `}, t)
	checkString("\xE2\x82\xAC", &Context{json: `"\u20AC" `}, t)
	checkString("你好吗", &Context{json: `"你好吗" `}, t)
}

func TestObject(t *testing.T) {

	json := `
{
"n" : null  ,
"f" : false ,
"t" : true,
"i" : 123,
"s" : "abc", 
"a" : [ 1, 2, 3],
"o" : { "1" : 1, "2" : 2, "3" : 3}
}
`
	value, err := Decode(json)
	if value == nil || err != nil {
		t.Fatal("err")
	}

	if value.(map[string]interface{})["n"] != nil {
		t.Fatal("err")
	}

	if value.(map[string]interface{})["f"].(bool) {
		t.Fatal("err")
	}

	if !value.(map[string]interface{})["t"].(bool) {
		t.Fatal("err")
	}

	v := value.(map[string]interface{})["i"].(float64)

	if !isDoubleEqual(v, 123) {
		t.Fatal("err")
	}

	if value.(map[string]interface{})["s"].(string) != "abc" {
		t.Fatal("err")
	}

	a := value.(map[string]interface{})["a"].([]interface{})

	for i := 0; i < 3; i++ {
		if !isDoubleEqual(a[i].(float64), float64(i+1)) {
			t.Fatal("err")
		}
	}

	o := value.(map[string]interface{})["o"].(map[string]interface{})

	if !isDoubleEqual(o["3"].(float64), 3) {
		t.Fatal("err")
	}

	if !isDoubleEqual(o["2"].(float64), 2) {
		t.Fatal("err")
	}

	if !isDoubleEqual(o["1"].(float64), 1) {
		t.Fatal("err")
	}

	fmt.Println(Encode(value))
}

func TestSimpleKind(t *testing.T) {
	json := `null`

	value, err := Decode(json)
	if value != nil || err != nil {
		t.Fatal("err")
	}

	json = `true`

	value, err = Decode(json)
	if value == nil || err != nil {
		t.Fatal("err")
	}

	if !value.(bool) {
		t.Fatal("err")
	}

	json = `"shit"`

	value, err = Decode(json)
	if value == nil || err != nil {
		t.Fatal("err")
	}

	if value.(string) != "shit" {
		t.Fatal("err")
	}

	json = `123456`

	value, err = Decode(json)
	if value == nil || err != nil {
		t.Fatal("err")
	}

	if !isDoubleEqual(value.(float64), 123456) {
		t.Fatal("err")
	}

	json = `[1,2,3]`

	value, err = Decode(json)
	if value == nil || err != nil {
		t.Fatal("err")
	}

	if !isDoubleEqual(value.([]interface{})[0].(float64), 1) {
		t.Fatal("err")
	}
}

File: json/javascript/json.test.js
let expect = require('chai').expect;
let _ = require('lodash');


let jsonParse = require('./json').jsonParse;

let isDoubleEqual = function (f1, f2) {
    return Math.abs(f1 - f2) < 0.000001;
};


describe("test number", function () {
    it('', function () {
        expect(isDoubleEqual(1.0, jsonParse('1 '))).to.be.equal(true);
        expect(isDoubleEqual(0.0, jsonParse("-0 "))).to.be.equal(true);
        expect(isDoubleEqual(0.0, jsonParse("-0.0 "))).to.be.equal(true);
        expect(isDoubleEqual(1, jsonParse("1 "))).to.be.equal(true);

        expect(isDoubleEqual(-1.0, jsonParse("-1 "))).to.be.equal(true);
        expect(isDoubleEqual(1.5, jsonParse("1.5 "))).to.be.equal(true);
        expect(isDoubleEqual(-1.5, jsonParse("-1.5 "))).to.be.equal(true);
        expect(isDoubleEqual(3.1416, jsonParse("3.1416 "))).to.be.equal(true);
        expect(isDoubleEqual(1E10, jsonParse("1E10 "))).to.be.equal(true);
        expect(isDoubleEqual(1e10, jsonParse("1e10 "))).to.be.equal(true);
        expect(isDoubleEqual(1E+10, jsonParse("1E+10 "))).to.be.equal(true);
        expect(isDoubleEqual(1E-10, jsonParse("1E-10 "))).to.be.equal(true);
        expect(isDoubleEqual(-1E10, jsonParse("-1E10 "))).to.be.equal(true);
        expect(isDoubleEqual(-1e10, jsonParse("-1e10 "))).to.be.equal(true);
        expect(isDoubleEqual(-1E+10, jsonParse("-1E+10 "))).to.be.equal(true);
        expect(isDoubleEqual(-1E-10, jsonParse("-1E-10 "))).to.be.equal(true);
        expect(isDoubleEqual(1.234E+10, jsonParse("1.234E+10 "))).to.be.equal(true);
        expect(isDoubleEqual(1.234E-10, jsonParse("1.234E-10 "))).to.be.equal(true);
        expect(isDoubleEqual(0.0, jsonParse("1e-10000 "))).to.be.equal(true);
        /* must underflow */

        expect(isDoubleEqual(1.0000000000000002, jsonParse("1.0000000000000002 "))).to.be.equal(true);
        /* the smallest number > 1 */
        expect(isDoubleEqual(4.9406564584124654e-324, jsonParse("4.9406564584124654e-324 "))).to.be.equal(true);
        /* minimum denormal */
        expect(isDoubleEqual(-4.9406564584124654e-324, jsonParse("-4.9406564584124654e-324 "))).to.be.equal(true);
        expect(isDoubleEqual(2.2250738585072009e-308, jsonParse("2.2250738585072009e-308 "))).to.be.equal(true);
        /* Max subnormal double */
        expect(isDoubleEqual(-2.2250738585072009e-308, jsonParse("-2.2250738585072009e-308 "))).to.be.equal(true);
        expect(isDoubleEqual(2.2250738585072014e-308, jsonParse("2.2250738585072014e-308 "))).to.be.equal(true);
        /* Min normal positive double */
        expect(isDoubleEqual(-2.2250738585072014e-308, jsonParse("-2.2250738585072014e-308 "))).to.be.equal(true);
        expect(isDoubleEqual(1.7976931348623157e+308, jsonParse("1.7976931348623157e+308 "))).to.be.equal(true);
        /* Max double */
        expect(isDoubleEqual(-1.7976931348623157e+308, jsonParse("-1.7976931348623157e+308 "))).to.be.equal(true);
    });
});

describe("test string", function () {
    it('', function () {
        expect(jsonParse('""')).to.be.equal("");
        expect(jsonParse('"Hello"')).to.be.equal("Hello");
        expect(jsonParse('"Hello\\nWorld"')).to.be.equal("Hello\nWorld");
        expect(jsonParse('"\\" \\\\ \\/ \\b \\f \\n \\r \\t"')).to.be.equal("\" \\ / \b \f \n \r \t");
        expect(jsonParse('"Hello\\u0000World"')).to.be.equal("Hello\u0000World");
        expect(jsonParse('"\\u0024"')).to.be.equal("\x24");
        //expect(jsonParse('"\\u20AC"')).to.be.equal("\xE2\x82\xAC");
        expect(jsonParse('"你好吗"')).to.be.equal("你好吗");
    });
});

describe("test object", function () {
    it('', function () {
        let json = `
 {
 "n" : null  ,
"f" : false ,
"t" : true,
"i" : 123,
"s" : "abc", 
"a" : [ 1, 2, 3],
"o" : { "1" : 1, "2" : 2, "3" : 3}
}`;

        let obj = jsonParse(json);
        expect(obj["n"] === null).to.be.equal(true);
        expect(obj["f"] === false).to.be.equal(true);
        expect(obj["t"] === true).to.be.equal(true);
        expect(isDoubleEqual(obj["i"], 123)).to.be.equal(true);
        expect(obj["s"] === "abc").to.be.equal(true);
        expect(isDoubleEqual(obj["a"][2], 3)).to.be.equal(true);
        expect(isDoubleEqual(obj["o"]["2"], 2)).to.be.equal(true);
    });
});

describe("test simple kind", function () {
    it('', function () {
        expect(jsonParse("null") === null).to.be.equal(true);
        expect(jsonParse("true") === true).to.be.equal(true);
        expect(jsonParse(`"shit"`) === "shit").to.be.equal(true);
        expect(isDoubleEqual(jsonParse(`123456`),123456)).to.be.equal(true);
        expect(isDoubleEqual(jsonParse(`[1,2,3]`)[0],1)).to.be.equal(true);
    });
});
